File: backend/app/schemas/ai_config.py
from pydantic import BaseModel, Field, field_validator


class AIConfigOut(BaseModel):
    """Runtime AI configuration exposed to the admin UI."""

    llm_provider: str = ""
    llm_api_key: str = ""  # masked when read from the backend
    llm_model: str = ""
    llm_base_url: str = ""
    llm_timeout: int = Field(default=5, ge=1, le=120)
    capa_draft_llm_timeout: int = Field(default=15, ge=1, le=120)
    report_llm_timeout: int = Field(default=10, ge=1, le=120)

    embedding_provider: str = ""
    embedding_model: str = ""
    embedding_base_url: str = ""
    embedding_dimensions: int = Field(default=1536, ge=1, le=4096)

    search_vector_weight: float = Field(default=0.7, ge=0, le=1)
    search_fulltext_weight: float = Field(default=0.3, ge=0, le=1)


class AIConfigUpdate(BaseModel):
    """Fields an admin can update from the UI."""

    llm_provider: str = ""
    llm_api_key: str = ""  # send empty to keep current; send new value to update
    llm_model: str = ""
    llm_base_url: str = ""
    llm_timeout: int = Field(default=5, ge=1, le=120)
    capa_draft_llm_timeout: int = Field(default=15, ge=1, le=120)
    report_llm_timeout: int = Field(default=10, ge=1, le=120)

    embedding_provider: str = ""
    embedding_model: str = ""
    embedding_base_url: str = ""
    embedding_dimensions: int = Field(default=1536, ge=1, le=4096)

    search_vector_weight: float = Field(default=0.7, ge=0, le=1)
    search_fulltext_weight: float = Field(default=0.3, ge=0, le=1)

    @field_validator("search_vector_weight", "search_fulltext_weight")
    @classmethod
    def _precision(cls, v: float) -> float:
        return round(v, 2)


class ProviderTestResultSchema(BaseModel):
    ok: bool
    latency_ms: int | None = None
    detail: str | None = None


class AIConfigTestResultSchema(BaseModel):
    llm: ProviderTestResultSchema
    embedding: ProviderTestResultSchema

File: frontend/src/api/aiConfig.ts
import client from "./client";

export interface AIConfig {
  llm_provider: string;
  llm_api_key: string;
  llm_model: string;
  llm_base_url: string;
  llm_timeout: number;
  capa_draft_llm_timeout: number;
  report_llm_timeout: number;
  embedding_provider: string;
  embedding_api_key: string;
  embedding_model: string;
  embedding_base_url: string;
  embedding_dimensions: number;
  search_vector_weight: number;
  search_fulltext_weight: number;
}

export async function getAIConfig(): Promise<AIConfig> {
  const { data } = await client.get("/admin/ai-config");
  return data;
}

export async function updateAIConfig(config: AIConfig): Promise<AIConfig> {
  const { data } = await client.put("/admin/ai-config", config);
  return data;
}

export interface ProviderTestResult {
  ok: boolean;
  latency_ms: number | null;
  detail: string | null;
}

export interface AIConfigTestResult {
  llm: ProviderTestResult;
  embedding: ProviderTestResult;
}

export async function testAIConfig(config: AIConfig): Promise<AIConfigTestResult> {
  const { data } = await client.post("/admin/ai-config/test", config);
  return data;
}
